<html>
<!-- home is the default webpage. we didnt yet decide will be shown here. suggestion can be search options/ example/
        summary on site/updates... 
-->

<head>
    <!-- load angular -->
    <script src="https://ajax.googleapis.com/ajax/libs/angularjs/1.7.8/angular.js"></script>
    <script src="https://code.angularjs.org/1.7.8/angular-route.js"></script>

    <!-- load app -->
    <script src="app.js"></script>

</head>
<!-- css -->
<style>
</style>


<!-- will be commented on when it will be created -->

<body ng-app="DoChaP" ng-controller="documentationCtrl">
        <div
        style="background-image: url('/resources/dna100.jpg'); height: 1000px;background-repeat: no-repeat; background-size: cover; background-position: center center;">
        <br>
        <br>
        <div class="searchItems" style="background-color:rgb(235, 235, 235); padding: 30px;  border-radius: 10px;width: 1200px">
                <p><strong>What does DoChaP do?</strong></p>
                <p>Alternative RNA splicing results in multiple transcripts of the same gene, possibly coding different protein isoforms that may differ in their protein domains. In order to study the connection between splice junctions and protein domains, researchers need to search for their gene of interest in several sources and integrate information of various formats to know how does alternative splicing affects the functional domains in the proteins.</p>
                <p>DoChaP provides fast, integrated and intuitive visualization to explore all the known splicing junctions that falls on protein domains.</p>
                <p>The representation of all the transcript of the gene might help researchers study their specific isoform or better interpret their differential splicing results.</p>
                <p>Sources:</p>
                <ul>
                <li>NCBI - RefSeq database (.gpff GeneBank files) as well as gene2ensemble table from Gene database</li>
                <li>UCSC - Tables (ncbiRefSeq, kgXref and knownGene) &ndash; of the latest genome versions for each species.</li>
                </ul>
                <p><strong>&nbsp;</strong></p>
                <p><strong>How to use DoChaP?</strong></p>
                <p>In the homepage of DoChaP you will find the search window where you need to provide your search term and choose species.</p>
                <p><em><u>A - Species: </u></em></p>
                <p>Human (hg38) or Mouse (mm10). More species will be added in the future. We invite users to contact us and ask for specific species of interest.</p>
                <p><em><u>B - Search by gene, transcript or proteins:</u></em></p>
                <p>In the search box, enter one of the following gene identifiers: gene symbol, gene synonym, NCBI&rsquo;s Gene ID, RefSeq transcript or protein id (starts with NM_ /XM_/NP_/XP_), ENSEMBL gene, transcript or protein id (starts with ENS[species prefix]G / ENS[species prefix]T / ENS[species prefix]P)</p>
                <p><em><u>C - Ignore predicted records:</u></em></p>
                <p>Predicted records are transcripts that were not yet validated by the providers of the source information, for example: transcripts or proteins with RefSeq id that starts with XM_ or XP_ (respectively). By checking this check box you will remove all the predictions from your results.</p>
                <p>&nbsp;</p>
               <img src="/resources/help1.png"/>
                <p><strong>&nbsp;</strong></p>
                <p><strong>How to interpret DoChaP results</strong><strong>?</strong></p>
                <p>After DoChaP found your gene of interest, it will present a page similar to Figure 2. For information, follow the description below:</p>
                <p><em><u>A &ndash; Title and display</u></em><em><u>: </u></em></p>
                <p>In the title you will see the number of isoforms, gene symbol and species.</p>
                <p>Below, you can adjust the display to present:</p>
                <ul>
                <li>All views (default) &ndash; Show gene structure (on the left) and the corresponding mRNA and protein domain composition (on the right), for all transcripts</li>
                <li>Genomic view only &ndash; Only show the gene structure for all transcripts</li>
                <li>Transcript view only - Only show the exons for all transcripts</li>
                <li>Protein view only &ndash; Only show the isoforms domain composition, for all transcripts</li>
                </ul>
                <p>Adjust the displayed genomic region by entering start or end position, overlapping the gene region.</p>
                <p><em><u>B - Genomic scale</u></em><em><u>:</u></em></p>
                <p>Left side: The black line represents the axis of genomic location with specific location ticks above. The currently displayed genomic location is shown above the genomic scale. Below the scale is an arrow representing coding strand (left arrow for reversed and right arrow for forward).</p>
                <p>Right side: The black line represents the axis of the transcript of the protein length, starting with zero, upper ticks represent nucleotide and lower ticks represent amino acids. All transcripts and proteins below are aligned to the zero.</p>
                <p><em><u>C &ndash; Genomic visualization of the gene&rsquo;s isoform</u></em><em><u>:</u></em></p>
                <p>The black thin line represents the genome, exons are marked in colored boxes. Untranslated regions are in thinner boxes at the edges of the transcript. Different colors represent different exons of the gene and the colors are consistent across transcript and protein visualization. Transcript RefSeq ID is shown on the left and clicking on it will show more information and external links. Hide transcript option will remove the transcript from the visualization.</p>
                <p><em><u>D &ndash; mRNA visualization of the gene&rsquo;s isoform</u></em><em><u>:</u></em></p>
                <p>The corresponding mature mRNA of the pre-mRNA on the left. All the exons (within the CDS) are joined, each color represents different exon within the transcript but consistent for all the transcripts of the gene. RefSeq transcript ID is shown on the right and clicking on it will show more information and external links.</p>
                <p><em><u>E- Protein domains composition visualization of the gene&rsquo;s isoform</u></em><em><u>:</u></em></p>
                <p>The corresponding protein isoform of the genomic representation on the left and the mRNA representation above. Black thin line represents the entire protein, domains are shown in round shapes and colored according to the exons they are composed of. The white lines represent splice junctions within the domain. RefSeq protein ID is shown on the right and clicking on it will show more information and external links.</p>
                <p>&nbsp;</p>
                <img src="/resources/help2.png"/>
                <p>&nbsp;</p>
                <p><strong>Compare species transcripts with DoChaP:</strong></p>
                <p>DoChaP "Compare Species" feature enable the user to visualize and compare transcripts from different species. To compare, choose species to compare and a gene of interest in a similar format to DoChaP search for one species, explained above.</p>
                <p>The results will be all the transcripts of the gene in species A followed by all the transcripts of the gene in species B, shown in the same format as described above for DoChaP main page.</p>
                <p>The color of the exons is consistent between the compared species, meaning that if the first exon of the transcript is the same size in the two species, then it will share the same color.</p>
                <p><strong>What is the source for the information DoChaP is showing?</strong></p>
                <p>DoChaP is integrating information from several sources, all downloaded from the relevant ftp sites:</p>
                <ul>
                <li>Exons annotations &ndash; parsed from <em>ncbiRefSeq</em> table downloaded from <a href="ftp://hgdownload.soe.ucsc.edu/goldenPath">UCSC database</a></li>
                <li>Protein domain annotations &ndash; parsed from GeneBank <em>.gpff </em>files from <a href="ftp://ftp.ncbi.nlm.nih.gov/refseq/">NCBI </a><a href="ftp://ftp.ncbi.nlm.nih.gov/refseq/">RefSeq</a><a href="ftp://ftp.ncbi.nlm.nih.gov/refseq/"> database</a></li>
                <li>Other tables used for accession id connections are: <em>kgXref</em> and <em>knownGene</em> tables from UCSC and <em>gene2ensemble</em> table from <a href="ftp://ftp.ncbi.nih.gov/gene/">NCBI Gene database</a></li>
                </ul>
                <p>All information was parsed and organized in one database for the use of DoChaP webserver.</p>
                <p>Of Note:</p>
                <ul>
                <li>Only protein coding genes are shown in DoChaP</li>
                <li>Domains are only the "Regions" features from NCBI RefSeq. Filtered for:
                <ul>
                <li>Records with accession id of known domain database (NCBI Curated Domains, Pfam, SMART, COG, PRK, TIGRFAMs)</li>
                <li>Records larger than 1 amino acid</li>
                </ul>
                </li>
                </ul>
                <p>&nbsp;</p>
                <p>&nbsp;</p>
                <p><strong>Browser Compatibility</strong></p>
                <table width="0">
                <tbody>
                <tr>
                <td width="86">
                <p>OS</p>
                </td>
                <td width="102">
                <p>Version</p>
                </td>
                <td width="71">
                <p>Chrome</p>
                </td>
                <td width="84">
                <p>Firefox</p>
                </td>
                <td width="115">
                <p>Microsoft Edge</p>
                </td>
                <td width="82">
                <p>Safari</p>
                </td>
                </tr>
                <tr>
                <td width="86">
                <p>Linux</p>
                </td>
                <td width="102">
                <p>Ubuntu 16</p>
                </td>
                <td width="71">79.0</td>
                <td width="84">71.0</td>
                <td width="115">
                <p>n/a</p>
                </td>
                <td width="82">
                <p>n/a</p>
                </td>
                </tr>
                <tr>
                <td width="86">
                <p>MacOS</p>
                </td>
                <td width="102">
                <p>Sierra 10.12.6</p>
                </td>
                <td width="71">
                <p>78.0</p>
                </td>
                <td width="84">
                <p>68.0.1</p>
                </td>
                <td width="115">
                <p>n/a</p>
                </td>
                <td width="82">
                <p>12.1.2</p>
                </td>
                </tr>
                <tr>
                <td width="86">
                <p>Windows</p>
                </td>
                <td width="102">
                <p>10</p>
                </td>
                <td width="71">
                <p>78.0</p>
                </td>
                <td width="84">
                <p>71.0</p>
                </td>
                <td width="115">
                <p>42.17134.1038.0</p>
                </td>
                <td width="82">
                <p>n/a</p>
                </td>
                </tr>
                </tbody>
                </table>
                <p>&nbsp;</p>
                <p>&nbsp;</p>
        </div>
    </div>

   
    
    
</body>

</html>