<html>
<!-- 
  here we explain and elaborate on the site's creations graphic and analyzed data.
 -->
<body ng-app="DoChaP" ng-controller="helpCtrl">
        <br>
        <div class="greyArea">
          <p><strong>How to use DoChaP?</strong></p>
          <p>On the homepage of DoChaP you will find the search window where you need to insert your search term and choose a species.</p>
          <p><em><u>A - Species: </u></em></p>
          <p>Human (hg38) or Mouse (mm10). More species will be added in the future. We invite users to contact us to request specific species of interest.</p>
          <p><em><u>B - Search by gene, transcript or protein:</u></em></p>
          <p>In the search box, enter one of the following gene identifiers: gene symbol, gene synonym, NCBI&rsquo;s Gene ID, RefSeq transcript or protein ID (starts with NM_ /XM_/NP_/XP_), ENSEMBL gene, transcript or protein ID (starts with ENS[species prefix]G / ENS[species prefix]T / ENS[species prefix]P)</p>
          <p><em><u>C - Ignore predicted records:</u></em></p>
          <p>Predicted records are transcripts that have not yet been validated by the providers of the source information; for example: transcripts or proteins with a RefSeq ID that starts with XM_ or XP_ (respectively). By checking this check box, you will remove all the predictions from your results.</p>
          <p>&nbsp;</p>
          <p><strong>How to interpret DoChaP results</strong><strong>?</strong></p>
          <p>After DoChaP has found your gene of interest, it will present a page similar to Figure 2. For information, follow the description below:</p>
          <p><em><u>&nbsp;</u></em></p>
          <p><em><u>A &ndash; Title and display:</u></em></p>
          <p>In the title, you will see the number of isoforms, the gene symbol and the species.</p>
          <p>The display below the title can be adjusted to present:</p>
          <ul>
          <li>All views (default) &ndash; Show gene structure (on the left) and the corresponding mRNA and the protein domain composition (on the right), for all transcripts</li>
          <li>Genomic view only &ndash; Show only the gene structure for all transcripts</li>
          <li>Transcript view only &ndash; Show only the exons for all transcripts</li>
          <li>Protein view only &ndash; Show only the domain composition of the isoforms for all transcripts</li>
          </ul>
          <p>Adjust the displayed genomic region by entering start or end position, overlapping the gene region.</p>
          <p><em><u>B &ndash; Genomic scale</u></em><em><u>:</u></em></p>
          <p>Left side: The black line represents the axis of the genomic location with specific location ticks above. The currently displayed genomic location is shown above the genomic scale. Below the scale is an arrow representing the coding strand (left arrow for reversed and right arrow for forward).</p>
          <p>Right side: The black line represents the axis of the transcript of the protein length, starting with zero. Upper ticks represent nucleotides, and lower ticks represent amino acids. All transcripts and proteins below are aligned to the zero.</p>
          <p><em><u>C &ndash; Genomic visualization of the gene&rsquo;s isoforms</u></em><em><u>:</u></em></p>
          <p>The black thin line represents the genome, and exons are marked in colored boxes. Untranslated regions are represented in thinner boxes at the edges of the transcript. Different colors represent different exons of the gene, and the colors are consistent across transcript and protein visualizations. Transcript RefSeq ID is shown on the left and clicking on it will show more information and external links. 'Hide transcript' option will remove the transcript from the visualization.</p>
          <p><em><u>D &ndash; mRNA visualization of the gene&rsquo;s isoforms:</u></em></p>
          <p>The mature mRNA that is edited from the pre-mRNA on the left. All the exons (within the coding region) are joined; each color represents a different exon within the transcript and is consistent across all the transcripts of the gene. The RefSeq transcript ID is shown on the right, and clicking on it will show more information and external links.</p>
          <p><em><u>E &ndash; Visualization of the protein domains composition of the gene&rsquo;s isoforms:</u></em></p>
          <p>The protein isoform of encoded from the genomic representation on the left and the mRNA representation above is shown. The thin black line represents the entire protein. Domains are shown in round shapes and are colored according to the exons they are composed of. White lines represent splice junctions within the domain. The RefSeq protein ID is shown on the right, and clicking on it will show more information and external links.</p>
          <p>&nbsp;</p>
          <p><strong>Compare species transcripts with DoChaP:</strong></p>
          <p>The DoChaP "Compare Species" feature enables the user to visualize and compare transcripts from different species. To compare transcripts, choose the species you wish to compare and a gene of interest in a similar format to the DoChaP search for a single species, as explained above.</p>
          <p>The results will be all the transcripts of the gene in species A followed by all the transcripts of the gene in species B, shown in the same format as described above for DoChaP main page.</p>
          <p><strong>&nbsp;</strong></p>
          <p><strong>What are the sources of the information in DoChaP?</strong></p>
          <p>DoChaP integrates data from NCBI&rsquo;s RefSeq, Ensembl and Interpro, into one combined database. Data sources:</p>
          <ul>
          <li>Genomic annotations (genes, exons, CDS, transcription) &ndash; .<em>gff</em> files downloaded from <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/">NCBI's genomes - RefSeq </a>and <a href="http://ftp.ensembl.org/pub/current_gff3/">Ensembl</a> ftp sites (the latest version available at the time of DoChaP update)</li>
          <li>Protein domain annotation &ndash; <em>.gpff </em>files were downloaded from the <a href="ftp://ftp.ncbi.nlm.nih.gov/refseq/">NCBI </a><a href="ftp://ftp.ncbi.nlm.nih.gov/refseq/">RefSeq</a><a href="ftp://ftp.ncbi.nlm.nih.gov/refseq/"> database</a> and Ensembl records were downloaded from <a href="https://www.ensembl.org/biomart/martview">BioMart</a>.</li>
          <li>Orthology data &ndash; downloaded from ensembl using <a href="https://www.ensembl.org/biomart/martview">BioMart</a>.</li>
          <li>Other tables used for accession ID connections are: <em>gene2ensemble</em> table from the <a href="ftp://ftp.ncbi.nih.gov/gene/">NCBI Gene database</a> and <em>InterPro entries </em>table downloaded from <a href="https://www.ebi.ac.uk/interpro/entry/InterPro/#table">InterPro</a>.</li>
          </ul>
          <p>All information was parsed and organized into a single database for the use of DoChaP webserver.</p>
          <p>Note:</p>
          <ul>
          <li>Only protein coding transcripts are shown in DoChaP</li>
          <li>Alternative chromosomes are not shown</li>
          <li>DoChaP includes domains from one of the following domain prediction sources: NCBI Curated Domains, Pfam, SMART, TIGRFAMs, Interpro</li>
          </ul>
          <p>&nbsp;</p>
          <p>DoChaP database update:</p>
          <ul>
          <li>The data in DoChaP is relevant to: May 25<sup>th</sup>, 2020 (last download from external sources)</li>
          <li>DoChaP last update: June 2<sup>nd</sup>, 2020</li>
          </ul>
          <p>&nbsp;</p>
          <p>&nbsp;</p>
          <p><strong>Browser Compatibility</strong></p>
          <table width="540">
          <tbody>
          <tr>
          <td width="86">
          <p>OS</p>
          </td>
          <td width="102">
          <p>Version</p>
          </td>
          <td width="71">
          <p>Chrome</p>
          </td>
          <td width="84">
          <p>Firefox</p>
          </td>
          <td width="115">
          <p>Microsoft Edge</p>
          </td>
          <td width="83">
          <p>Safari</p>
          </td>
          </tr>
          <tr>
          <td width="86">
          <p>Linux</p>
          </td>
          <td width="102">
          <p>Ubunto 16</p>
          </td>
          <td width="71">
          <p>79.0</p>
          </td>
          <td width="84">
          <p>71.0</p>
          </td>
          <td width="115">
          <p>n/a</p>
          </td>
          <td width="83">
          <p>n/a</p>
          </td>
          </tr>
          <tr>
          <td width="86">
          <p>MacOS</p>
          </td>
          <td width="102">
          <p>Sierra 10.12.6</p>
          </td>
          <td width="71">
          <p>78.0</p>
          </td>
          <td width="84">
          <p>68.0.1</p>
          </td>
          <td width="115">
          <p>n/a</p>
          </td>
          <td width="83">
          <p>12.1.2</p>
          </td>
          </tr>
          <tr>
          <td width="86">
          <p>Windows</p>
          </td>
          <td width="102">
          <p>10</p>
          </td>
          <td width="71">
          <p>78.0</p>
          </td>
          <td width="84">
          <p>71.0</p>
          </td>
          <td width="115">
          <p>42.17134</p>
          </td>
          <td width="83">
          <p>n/a</p>
          </td>
          </tr>
          </tbody>
          </table>
          <p>&nbsp;</p>
          <p>&nbsp;</p>
        </div>
</body>

</html>