<html>
<!-- 
  here we explain and elaborate on the site's creations graphic and analyzed data.
 -->
<body ng-app="DoChaP" ng-controller="helpCtrl">
        <br>
        <div class="greyArea">
          <p><strong>How to use DoChaP?</strong></p>
<p>On the homepage of DoChaP you will find the search window where you can insert your search term and choose a species.</p>
<p><img src="resources/help_guide_fig1.png"  width="600"></p>
<p><em><u>A - Species: </u></em></p>
<p>Select one of the available species. Five species are currently available: Human (hg38), &nbsp;Mouse (mm10), Rat (rn6), Zebrafish (danRer11), Frog (xenTro9).</p>
<p>More species might be added in the future. We invite users to <a href="#!contactUs">contact us</a> to request specific species of interest.</p>
<p><em><u>B - Search by gene, transcript or protein:</u></em></p>
<p>In the search box, type your gene of interest using one of the following identifiers: gene symbol, gene synonym, NCBI&rsquo;s Gene ID, RefSeq transcript or protein ID (starts with NM_ /XM_/NP_/XP_), ENSEMBL gene, transcript or protein ID (starts with ENS).</p>
<p>Finally, click the &ldquo;Search&rdquo; button. Note, for genes with many transcript the search might take a while.</p>
<p>Below, you can find some examples for potential queries.</p>
<p><strong>How to interpret DoChaP results</strong><strong>?</strong></p>
<p>After DoChaP has found your gene of interest, it will present a page similar to Figure 2. For information, follow the description below:</p>
<p><img src="resources/help_guide_fig2.png"  width="900"></p>
<p><em><u>&nbsp;</u></em></p>
<p><em><u>A &ndash; Title and display:</u></em></p>
<p>The title includes the number of isoforms, the gene symbol and the species and genome version presented in the results page.</p>
<p>The display below the title can be adjusted using the &ldquo;Choose Display&rdquo; to present:</p>
<ul>
<li>All views (default) &ndash; Show the genomic structure of the transcript (on the left) and the corresponding coding region of the transcript and the protein domain composition (on the right), for all transcripts</li>
<li>Transcript &amp; Protein views &ndash; Show the coding transcript and the protein domain composition (similar to the right side of the &ldquo;All views&rdquo; option), for all transcripts</li>
<li>Genomic view only &ndash; Shows only the transcript structure for all transcripts</li>
<li>Transcript view only &ndash; Shows only the coding regions of all transcripts colored by exons</li>
<li>Protein view only &ndash; Shows only the protein domain composition of the isoforms for all transcripts</li>
</ul>
<p>&ldquo;Hide predicted records&rdquo;</p>
<ul>
<li>Predicted records are transcripts that have not yet been validated by the providers of the source information; for example: transcripts or proteins with a RefSeq ID that starts with XM_ or XP_ (respectively). By checking this check box, you will hide all the predictions from your results.</li>
</ul>
<p>&ldquo;Save as PDF&rdquo;</p>
<ul>
<li>Use this option to save the results of your gene of interest to a high quality PDF file.</li>
</ul>
<p><em><u>&nbsp;</u></em></p>
<p><em><u>&nbsp;</u></em></p>
<p><em><u>B &ndash;Zoom in and sliding window</u></em></p>
<p>Zoom in panel to control the genomic coordinates (left side) and the transcribed region coordinates (right side) which are shown. The black line represents the shown region and the flags on the edges represent the start and end location of the shown window. The start and end location can be moved by the user and the black line can be slided to adjust the visualization.</p>
<p>The loci of the gene of interest is shown above the genomic zoom in panel.</p>
<p><em><u>C- Location bar</u></em></p>
<p>Left side: The black bar represents the axis of the genomic location with specific location ticks above. Below the bar is an arrow representing the coding strand (left arrow for reverse and right arrow for forward). Note, when genes encoded on the reverse strand are shown, the axis is flipped to allow the first exon to be on the left and last exon to be on the right, in order to be consistent with the transcript and protein visualization.</p>
<p>Right side: The black bar represents the axis of the transcript of the protein length, starting with zero. Upper ticks represent nucleotides, and lower ticks represent amino acids. All transcripts and proteins below are aligned to the zero. Only the coding region of the transcripts is shown (no introns and no UTR)</p>
<p><em><u>D &ndash; Genomic visualization of the gene&rsquo;s isoforms</u></em><em><u>:</u></em></p>
<p>The black thin line represents the genome, and exons are marked in colored boxes. Untranslated regions are represented in thinner boxes at the edges of the transcript. Different colors represent different exons of the gene, and the colors are consistent across transcript and protein visualizations. Transcript ID is shown above each transcript (RefSeq and Ensembl if exist or one of them). Clicking on it will show more information on the transcript and some external links. 'Hide transcript' option (<button class="btn blackButton"><span class="glyphicon glyphicon-eye-close"></span></button> ) will hide the transcript from the visualization. Hovering over the exons will show the genomic position of the exon and its ordinal number in the transcript.</p>
<p><em><u>E &ndash; Coding mRNA visualization of the gene&rsquo;s isoforms:</u></em></p>
<p>The coding region of mature mRNA that is edited from the pre-mRNA, without UTRs. All the exons (within the coding region) are joined; each color represents a different exon within the transcript and is consistent across all the transcripts of the same gene. The transcript ID is shown above (RefSeq and Ensembl if exist or one of them), and clicking on it will show more information and external links. Hovering over the exons will show the ordinal number of the exon in the transcript and its length.</p>
<p>&nbsp;</p>
<p><em><u>F &ndash; Visualization of the protein domains composition of the isoforms of the gene:</u></em></p>
<p>The protein isoform encoded from the genomic structure of the transcript on the left and the corresponding coding region shown above. The thin black line represents the entire protein. Domains are shown in round shapes and are colored according to the exons encoding for them. White vertical lines represent splice junctions within the domain. The protein ID is displayed above the domain visualization, and clicking on it will show more information and external links. Hovering over a domain will show its full name, position on the protein, length in amino acids, external ID and a link to the external source (Pfam, SMART, CDD, TIGRfams, InterPro).</p>
<p>In the case of overlapping domains, a single domain with double frame is displayed. Hovering will show the number of overlapping domains, and clicking will open a detailed visualization showing all the domains (Figure 3).</p>
<p><img src="resources/help_guide_fig3.png" width="600"></p>
<p>&nbsp;</p>
<p><em><u>G- Scroll down for more&hellip;</u></em></p>
<p>More transcripts (if exist) are shown below, scroll down to see all. At the bottom of the page there is the exon color legend, indicating for each exon color its genomic location and transcripts containing this exon.</p>
<p>&nbsp;</p>
<p><strong>Compare species transcripts with DoChaP:</strong></p>
<p>The DoChaP "Compare Species" feature enables the user to visualize and compare orthologues genes from different species.</p>
<p><img src="resources/help_guide_fig4.png" width="900"></p>
<p>To compare transcripts, choose the first species you wish to compare and a gene of interest in a similar format to the DoChaP search for a single species (explained above), then click on &ldquo;Find orthology&rdquo;. DoChaP will look for all the orthologues of your gene in the other available species. Then select the species you want to compare to and the gene to compare from the given options. Orthologues are defined by Ensembl Compara. Genes that were not found in Ensembl Compara but share the same gene symbol in different species are also available to view.</p>
<p>The results will be all the transcripts of the gene in the first species on the left and all the transcripts of the gene in the second species on the right, shown in the same format as described above for DoChaP main page.</p>
<p><strong>&nbsp;</strong></p>
<p><strong>What are the sources of the information in DoChaP?</strong></p>
<p>DoChaP integrates data from multiple sources, see <a href="#!about">About</a> page for more information.</p>
<p>&nbsp;</p>
<p><strong>Browser Compatibility</strong></p>
<table width="540">
<tbody>
<tr>
<td width="86">
<p>OS</p>
</td>
<td width="102">
<p>Version</p>
</td>
<td width="71">
<p>Chrome</p>
</td>
<td width="84">
<p>Firefox</p>
</td>
<td width="115">
<p>Microsoft Edge</p>
</td>
<td width="83">
<p>Safari</p>
</td>
</tr>
<tr>
<td width="86">
<p>Linux</p>
</td>
<td width="102">
<p>Ubunto 16</p>
</td>
<td width="71">
<p>79.0</p>
</td>
<td width="84">
<p>71.0</p>
</td>
<td width="115">
<p>n/a</p>
</td>
<td width="83">
<p>n/a</p>
</td>
</tr>
<tr>
<td width="86">
<p>MacOS</p>
</td>
<td width="102">
<p>Sierra 10.12.6</p>
</td>
<td width="71">
<p>78.0</p>
</td>
<td width="84">
<p>68.0.1</p>
</td>
<td width="115">
<p>n/a</p>
</td>
<td width="83">
<p>12.1.2</p>
</td>
</tr>
<tr>
<td width="86">
<p>Windows</p>
</td>
<td width="102">
<p>10</p>
</td>
<td width="71">
<p>78.0</p>
</td>
<td width="84">
<p>71.0</p>
</td>
<td width="115">
<p>42.17134</p>
</td>
<td width="83">
<p>n/a</p>
</td>
</tr>
</tbody>
</table>

</body>

</html>