<html>
<!-- home is the default webpage. we didnt yet decide will be shown here. suggestion can be search options/ example/
        summary on site/updates... 
-->

<head>
    <!-- load angular -->
    <script src="https://ajax.googleapis.com/ajax/libs/angularjs/1.7.8/angular.js"></script>
    <script src="https://code.angularjs.org/1.7.8/angular-route.js"></script>

    <!-- load app -->
    <script src="app.js"></script>

</head>
<!-- css -->
<style>
</style>


<!-- will be commented on when it will be created -->

<body ng-app="DoChaP" ng-controller="documentationCtrl">

    <div
        style="background-image: url('/resources/dna100.jpg'); height: 1000px;background-repeat: no-repeat; background-size: cover; background-position: center center;">
        <br>
        <br>
        <div class="searchItems"
            style="background-color:rgb(235, 235, 235); padding: 30px;  border-radius: 10px;width: 900px">
            <p>The Domain Change Predictor (DoChaP) is a webserver for visualization of exonâ€“domain relationships in
                different transcripts. DoChaP provides an easy way of exploring the effect of alternative splicing on
                protein domains.</p>
            <p>Alternative RNA splicing results in multiple transcripts of the same gene, possibly coding different
                protein isoforms that may differ in their protein domains. Although transcripts and domains information
                can be extracted from public databases, such a process requires the integration of data from several
                sources and it is tedious and unfeasible for genome-wide analysis. DoChaP integrates transcripts and
                domains information to facilitate the study of alternative splicing effects on domain structure.
                Identifying such an effect can lead to testable hypotheses in the form of 'protein X loses domain Y in
                state Z'. <strong>&nbsp;</strong></p>
            <p>DoChaP integrates information on transcripts and protein domains data from the NCBI and UCSC databases
                into a single database.</p>
            <p>Sources:</p>
            <ul>
                <li>NCBI - RefSeq database (.gpff GeneBank files) and the gene2ensemble table from Gene database</li>
                <li>UCSC - Tables (ncbiRefSeq, kgXref and knownGene) of the latest genome versions for each species.
                </li>
            </ul>
            <p><strong>&nbsp;</strong></p>
            <p><strong>Developers:</strong> Shani T. Gal-Oz, Dana Eliyahu and Nimrod Haiat, from the
                <a href="http://lifeserv.bgu.ac.il/wp/talshay/">group of Tal Shay</a>, (Department of Life Sciences,
                Ben-Gurion University of the Negev), in collaboration with
                <a href="http://www.bgu.ac.il/~shanigu/">Guy Shani</a> (Department of Information Systems Engineering,
                Ben-Gurion University of the Negev, Israel).
            </p>
            <p>Contact: Shani T. Gal-Oz, <a href="mailto:galozs@post.bgu.ac.il" target="_blank">galozs@post.bgu.ac.il</a></p>
        </div>
    </div> 
</body>

</html>